File: android/app/src/main/kotlin/com/agent/aios/domain/agent/ResponseParser.kt
package com.agent.aios.domain.agent

import android.util.Log

class ResponseParser(private val validToolNames: Set<String>) {
    private val TAG = "AIOS-Parser"

    companion object {
        private val LINE_START_ACTION_REGEX = Regex("""(?im)^Action\s*:\s*(\w+)""")
        private val ARGS_REGEX = Regex("""(?im)^Args\s*:\s*(.+)""", RegexOption.MULTILINE)
        private val ANSWER_REGEX = Regex("""(?im)^Answer\s*:\s*(.+)""", RegexOption.DOT_MATCHES_ALL)
    }

    fun parse(response: String): ParseResult {
        val trimmed = response.trim()
        if (trimmed.isBlank()) return ParseResult.Empty

        try {
            val actionMatch = LINE_START_ACTION_REGEX.find(trimmed)
            if (actionMatch != null) {
                val toolName = actionMatch.groupValues[1].lowercase()
                if (toolName in validToolNames) {
                    val afterAction = trimmed.substring(actionMatch.range.last + 1)
                    val argsStart = afterAction.indexOf('{')
                    val args =
                        if (argsStart >= 0) {
                            extractJsonArgs(afterAction, argsStart)
                        } else {
                            ARGS_REGEX.find(trimmed)?.groupValues?.get(1)?.trim() ?: "{}"
                        }
                    return ParseResult.Action(toolName, args)
                }
            }

            ANSWER_REGEX.find(trimmed)?.let { match ->
                return ParseResult.Answer(match.groupValues[1].trim())
            }
        } catch (e: Exception) {
            Log.e(TAG, "parseResponse error: ${e.message}", e)
        }

        return ParseResult.Empty
    }

    private fun extractJsonArgs(
        text: String,
        startIndex: Int,
    ): String {
        var depth = 0
        var inString = false
        var escape = false
        for (i in startIndex until text.length) {
            val c = text[i]
            if (escape) {
                escape = false
                continue
            }
            if (c == '\\') {
                escape = true
                continue
            }
            if (c == '"') {
                inString = !inString
                continue
            }
            if (!inString) {
                if (c == '{') depth++
                if (c == '}') {
                    depth--
                    if (depth == 0) return text.substring(startIndex, i + 1)
                }
            }
        }
        return text.substring(startIndex)
    }

    sealed class ParseResult {
        data class Action(val toolName: String, val args: String) : ParseResult()

        data class Answer(val text: String) : ParseResult()

        data object Empty : ParseResult()
    }
}

File: lib/domain/agent/response_parser.dart
class ResponseParser {
  ResponseParser(this.validToolNames);

  final Set<String> validToolNames;

  static final _actionRegex = RegExp(
    r'Action\s*:\s*(\w+)',
    multiLine: true,
    caseSensitive: false,
  );
  static final _argsRegex = RegExp(
    r'Args\s*:\s*(.+)',
    multiLine: true,
    caseSensitive: false,
  );
  static final _answerRegex = RegExp(
    r'Answer\s*:\s*(.+)',
    multiLine: true,
    caseSensitive: false,
    dotAll: true,
  );

  ParseResult parseIntent(String response) {
    final lower = response.trim().toLowerCase();
    if (lower.contains('conversation')) {
      return const ParseIntent(true);
    }
    if (lower.contains('task')) {
      return const ParseIntent(false);
    }
    return const ParseIntent(false);
  }

  ParseResult parse(String response) {
    final trimmed = response.trim();
    if (trimmed.isEmpty) return const ParseEmpty();

    try {
      final actionMatch = _actionRegex.firstMatch(trimmed);
      if (actionMatch != null) {
        final toolName = actionMatch.group(1)!.toLowerCase();
        if (validToolNames.contains(toolName)) {
          final afterAction = trimmed.substring(actionMatch.end);
          final argsStart = afterAction.indexOf('{');
          final args = argsStart >= 0
              ? _extractJsonArgs(afterAction, argsStart)
              : _argsRegex
                      .firstMatch(trimmed)
                      ?.group(1)
                      ?.trim() ??
                  '{}';
          return ParseAction(toolName, args);
        }
      }

      final answerMatch = _answerRegex.firstMatch(trimmed);
      if (answerMatch != null) {
        return ParseAnswer(answerMatch.group(1)!.trim());
      }
    } on Object catch (e) {
      print('[AIOS-Parser] ERROR: parseResponse error: $e');
    }

    return const ParseEmpty();
  }

  String _extractJsonArgs(String text, int startIndex) {
    var depth = 0;
    var inString = false;
    var escape = false;

    for (var i = startIndex; i < text.length; i++) {
      final c = text[i];
      if (escape) {
        escape = false;
        continue;
      }
      if (c == '\\') {
        escape = true;
        continue;
      }
      if (c == '"') {
        inString = !inString;
        continue;
      }
      if (!inString) {
        if (c == '{') depth++;
        if (c == '}') {
          depth--;
          if (depth == 0) return text.substring(startIndex, i + 1);
        }
      }
    }
    return text.substring(startIndex);
  }
}

sealed class ParseResult {
  const ParseResult();
}

class ParseAction extends ParseResult {
  final String toolName;
  final String args;
  const ParseAction(this.toolName, this.args);
}

class ParseAnswer extends ParseResult {
  final String text;
  const ParseAnswer(this.text);
}

class ParseEmpty extends ParseResult {
  const ParseEmpty();
}

class ParseIntent extends ParseResult {
  final bool isConversation;
  const ParseIntent(this.isConversation);
}
